eclair-core/src/test/scala/fr/acinq/eclair/payment/relay/ChannelRelayerSpec.scala

```diff
@@ -0,0 +1,423 @@
+      channelRelayer ! WrappedLocalChannelUpdate(u)
+      channelRelayer ! WrappedChannelRelayPacket(r)
+      val fwd = expectFwdAdd(register, shortId1, outgoingAmount, outgoingExpiry)
+      fwd.message.replyTo ! RES_ADD_FAILED(fwd.message, testCase.exc, Some(testCase.update))
+      expectFwdFail(register, r.add.channelId, CMD_FAIL_HTLC(r.add.id, Right(testCase.failure), commit = true))
+    }
+  }
+
+  test("settlement failure") { f =>
+    import f._
+
+    val channelId1 = channelIds(shortId1)
+    val payload = RelayLegacyPayload(shortId1, outgoingAmount, outgoingExpiry)
+    val r = createValidIncomingPacket(1100000 msat, CltvExpiry(400100), payload)
+    val u = createChannelUpdate(shortId1)
+    val u_disabled = createChannelUpdate(shortId1, enabled = false)
+    val downstream_htlc = UpdateAddHtlc(channelId1, 7, outgoingAmount, paymentHash, outgoingExpiry, emptyOnionPacket)
+
+    case class TestCase(result: HtlcResult, cmd: Command with HasHtlcId)
+
+    val testCases = Seq(
+      TestCase(HtlcResult.RemoteFail(UpdateFailHtlc(channelId1, downstream_htlc.id, hex"deadbeef")), CMD_FAIL_HTLC(r.add.id, Left(hex"deadbeef"), commit = true)),
+      TestCase(HtlcResult.RemoteFailMalformed(UpdateFailMalformedHtlc(channelId1, downstream_htlc.id, ByteVector32.One, FailureMessageCodecs.BADONION)), CMD_FAIL_MALFORMED_HTLC(r.add.id, ByteVector32.One, FailureMessageCodecs.BADONION, commit = true)),
+      TestCase(HtlcResult.OnChainFail(HtlcOverriddenByLocalCommit(channelId1, downstream_htlc)), CMD_FAIL_HTLC(r.add.id, Right(PermanentChannelFailure), commit = true)),
+      TestCase(HtlcResult.Disconnected(u_disabled.channelUpdate), CMD_FAIL_HTLC(r.add.id, Right(TemporaryChannelFailure(u_disabled.channelUpdate)), commit = true))
+    )
+
+    testCases.foreach { testCase =>
+      channelRelayer ! WrappedLocalChannelUpdate(u)
+      channelRelayer ! WrappedChannelRelayPacket(r)
+      val fwd = expectFwdAdd(register, shortId1, outgoingAmount, outgoingExpiry)
+      fwd.message.replyTo ! RES_SUCCESS(fwd.message, channelId1)
+      fwd.message.origin.replyTo ! RES_ADD_SETTLED(fwd.message.origin, downstream_htlc, testCase.result)
+      expectFwdFail(register, r.add.channelId, testCase.cmd)
+    }
+  }
+
+  test("settlement success") { f =>
+    import f._
+    val eventListener = TestProbe[ChannelPaymentRelayed]()
+    system.eventStream ! EventStream.Subscribe(eventListener.ref)
+
+    val channelId1 = channelIds(shortId1)
+    val payload = RelayLegacyPayload(shortId1, outgoingAmount, outgoingExpiry)
+    val r = createValidIncomingPacket(1100000 msat, CltvExpiry(400100), payload)
+    val u = createChannelUpdate(shortId1)
+    val downstream_htlc = UpdateAddHtlc(channelId1, 7, outgoingAmount, paymentHash, outgoingExpiry, emptyOnionPacket)
+
+    case class TestCase(result: HtlcResult)
+
+    val testCases = Seq(
+      TestCase(HtlcResult.RemoteFulfill(UpdateFulfillHtlc(channelId1, downstream_htlc.id, paymentPreimage)))
```

Comment:
Can't we also test the `OnChainFulfill`?

Reply:
Done with 9adec95c9.